drivers/adc/adc_esp32.c

```diff
@@ -0,0 +1,334 @@
+	uint8_t channel_count;
+};
+
+struct adc_esp32_data {
+	adc_atten_t attenuation[ADC_CHANNEL_MAX];
+	uint8_t resolution[ADC_CHANNEL_MAX];
+	esp_adc_cal_characteristics_t chars[ADC_CHANNEL_MAX];
+	uint16_t meas_ref_internal;
+	uint16_t *buffer;
+	uint16_t *buffer_repeat;
+	bool calibrate;
+};
+
+/* Convert zephyr,gain property to the ESP32 attenuation */
+static inline int gain2attenuation(enum adc_gain gain, adc_atten_t *atten)
+{
+	switch (gain) {
+	case ADC_GAIN_1:
+		*atten = ADC_ATTEN_DB_0;
+		break;
+	case ADC_GAIN_4_5:
+		*atten = ADC_ATTEN_DB_2_5;
+		break;
+	case ADC_GAIN_1_2:
+		*atten = ADC_ATTEN_DB_6;
+		break;
+	case ADC_GAIN_1_4:
+		*atten = ADC_ATTEN_DB_11;
+		break;
+	default:
+		return -1;
+	}
+	return 0;
+}
+
+/* Convert voltage by inverted attenuation to support zephyr gain values */
+static void attenuation2gain_invert(adc_atten_t atten, uint32_t *val_mv)
+{
+	if (!val_mv) {
+		return;
+	}
+	switch (atten) {
+	case ADC_ATTEN_DB_2_5:
+		*val_mv = (*val_mv * 4) / 5; /* 1/ADC_GAIN_4_5 */
+		break;
+	case ADC_ATTEN_DB_6:
+		*val_mv = *val_mv >> 1; /* 1/ADC_GAIN_1_2 */
+		break;
+	case ADC_ATTEN_DB_11:
+		*val_mv = *val_mv / 4; /* 1/ADC_GAIN_1_4 */
+		break;
+	case ADC_ATTEN_DB_0: /* 1/ADC_GAIN_1 */
+	default:
+		break;
+	}
+}
+
+static bool adc_calibration_init(const struct device *dev)
+{
+	struct adc_esp32_data *data = dev->data;
+
+	switch (esp_adc_cal_check_efuse(ADC_CALI_SCHEME)) {
+	case ESP_ERR_NOT_SUPPORTED:
+		LOG_WRN("Skip software calibration - Not supported!");
+		break;
+	case ESP_ERR_INVALID_VERSION:
+		LOG_WRN("Skip software calibration - Invalid version!");
+		break;
+	case ESP_OK:
+		LOG_DBG("Software calibration possible");
+		return true;
+	default:
+		LOG_ERR("Invalid arg");
+		break;
+	}
+	return false;
+}
+
+static int adc_esp32_read(const struct device *dev, const struct adc_sequence *seq)
+{
+	const struct adc_esp32_conf *conf = dev->config;
+	struct adc_esp32_data *data = dev->data;
+	int reading;
+	uint32_t cal, cal_mv;
+
+	uint8_t channel_id = find_lsb_set(seq->channels) - 1;
+
+	if (seq->buffer_size < 2) {
+		LOG_ERR("Sequence buffer space too low '%d'", seq->buffer_size);
+		return -ENOMEM;
+	}
+
+	if (seq->channels > BIT(channel_id)) {
+		LOG_ERR("Multi-channel readings not supported");
+		return -ENOTSUP;
+	}
+
+	if (INVALID_RESOLUTION(seq->resolution)) {
+		LOG_ERR("unsupported resolution (%d)", seq->resolution);
+		return -ENOTSUP;
+	}
+
+	if (seq->calibrate) {
+		/* TODO: Does this mean actual Vref measurement on selected GPIO ?*/
+		LOG_ERR("calibration is not supported");
+		return -ENOTSUP;
+	}
+
+	data->resolution[channel_id] = seq->resolution;
+
+#if CONFIG_IDF_TARGET_ESP32C3
+	/* NOTE: nothing to set on ESP32C3 SoC */
+	if (conf->unit == ADC_UNIT_1) {
+		adc1_config_width(ADC_WIDTH_BIT_DEFAULT);
+	}
+#else
+	adc_set_data_width(conf->unit, WIDTH_MASK(data->resolution[channel_id]));
+#endif
+
+	/* Read raw value */
+	if (conf->unit == ADC_UNIT_1) {
+		reading = adc1_get_raw(channel_id);
```

Comment:
It seems this reading mechanism is not thread safe, since the adc register polling is not atomic, and there is no other synchronization method for reentrant calls of this function.

Also I'm afraid of locking here in case of a high prio task, may result in a large waste of CPU depending on the ADC conversion rate

Reply:
`adcX_get_raw` is using interrupt locking